File: app/src/main/java/com/cimdriver/app/util/WorkHoursNormalizer.kt
package com.cimdriver.app.util

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.math.abs

object WorkHoursNormalizer {
    const val DEFAULT_TOLERANCE_MINUTES = 90L

    fun normalize(
        actualTime: Long,
        dayStart: Long,
        configuredTime: String,
        toleranceMinutes: Long = DEFAULT_TOLERANCE_MINUTES
    ): Long {
        val configuredTimestamp = configuredTimestamp(dayStart, configuredTime) ?: return actualTime
        val differenceMinutes = abs(actualTime - configuredTimestamp) / 60_000L
        return if (differenceMinutes <= toleranceMinutes) configuredTimestamp else actualTime
    }

    private fun configuredTimestamp(dayStart: Long, configuredTime: String): Long? {
        return try {
            val normalizedTime = configuredTime.replace(".", ":")
            val format = SimpleDateFormat("HH:mm", Locale.ROOT).apply { isLenient = false }
            val parsed = format.parse(normalizedTime) ?: return null
            val calendar = java.util.Calendar.getInstance().apply {
                timeInMillis = dayStart
                val time = java.util.Calendar.getInstance().apply { time = parsed }
                set(java.util.Calendar.HOUR_OF_DAY, time.get(java.util.Calendar.HOUR_OF_DAY))
                set(java.util.Calendar.MINUTE, time.get(java.util.Calendar.MINUTE))
                set(java.util.Calendar.SECOND, 0)
                set(java.util.Calendar.MILLISECOND, 0)
            }
            calendar.timeInMillis
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/CIMDriver/Utilities/WorkHoursNormalizer.swift
import Foundation

enum WorkHoursNormalizer {
    static let defaultToleranceMinutes: TimeInterval = 90.0

    static func normalize(
        actualTime: Date,
        dayStart: Date,
        configuredTime: String,
        toleranceMinutes: TimeInterval = defaultToleranceMinutes
    ) -> Date {
        guard let configuredTimestamp = configuredTimestamp(dayStart: dayStart, configuredTime: configuredTime) else {
            return actualTime
        }
        
        let differenceMinutes = abs(actualTime.timeIntervalSince(configuredTimestamp)) / 60.0
        
        if differenceMinutes <= toleranceMinutes {
            return configuredTimestamp
        }
        
        return actualTime
    }
    
    static func roundToNearestQuarterHour(_ date: Date) -> Date {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.year, .month, .day, .hour, .minute], from: date)
        let minute = components.minute ?? 0
        let roundedMinute = Int((Double(minute) / 15.0).rounded()) * 15
        
        var updated = components
        updated.minute = roundedMinute == 60 ? 0 : roundedMinute
        updated.second = 0
        updated.nanosecond = 0
        
        let baseDate = calendar.date(from: updated) ?? date
        return roundedMinute == 60 ? calendar.date(byAdding: .hour, value: 1, to: baseDate) ?? baseDate : baseDate
    }
    
    static func effectiveBreakMinutes(
        start: Date,
        end: Date,
        configuredBreakMinutes: Int,
        toleranceMinutes: Int
    ) -> Int {
        let durationMinutes = max(0, Int(end.timeIntervalSince(start) / 60.0))
        guard durationMinutes > 0 else { return 0 }
        
        let normalizedBreak = max(0, configuredBreakMinutes)
        if durationMinutes <= toleranceMinutes {
            return 0
        }
        
        return min(normalizedBreak, durationMinutes)
    }

    private static func configuredTimestamp(dayStart: Date, configuredTime: String) -> Date? {
        let normalizedTime = configuredTime.replacingOccurrences(of: ".", with: ":")
        
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        
        guard let parsed = formatter.date(from: normalizedTime) else { return nil }
        
        let calendar = Calendar.current
        
        let timeComponents = calendar.dateComponents([.hour, .minute], from: parsed)
        var dayComponents = calendar.dateComponents([.year, .month, .day], from: dayStart)
        
        dayComponents.hour = timeComponents.hour
        dayComponents.minute = timeComponents.minute
        dayComponents.second = 0
        dayComponents.nanosecond = 0
        
        return calendar.date(from: dayComponents)
    }
}
